sdk-actors/src/main/java/io/dapr/actors/runtime/DaprStateAsyncProvider.java

```diff
@@ -69,7 +73,7 @@
         }
 
         T response = this.stateSerializer.deserialize(s, type);
-        if (this.isStateSerializerDefault && (response instanceof byte[])) {
+        if (this.isStateSerializerJson && (response instanceof byte[])) {

```

Comment:
Changing this logic will make this not backwards compatible - it means that values saving with the current version of the code will not be able to be read from a newer version.

After thinking about this same problem in another PR, I think this should be done by a configuration. When registering the ActorType, a config can be set to determine is the data is written as Text or binary (Base64). The default behavior is Base64. This way, existing code does not break.

Reply:
Hi! I understood your point. But we have a unit test that checks that binary processing is OK. See https://github.com/dapr/java-sdk/blob/master/sdk-actors/src/test/java/io/dapr/actors/runtime/DaprStateAsyncProviderTest.java#L166
I add this test for a custom serializer and it works well. 
I can add a method to the serializer interface with the name `useMimeType` or `checkMimeType` and change behavior use this method

Reply:
@artursouza since Java 8 we have a `default` method. Because of that, it won't be breaking changes. I still can't imagine a case in which current changes will be breaking. Did you find this case with integration tests? May be beta tester give this information to you? 
I have a real system with a custom serializer and many states stored in binary format. I try changed SDK for binary format. All works as expected.

Reply:
@abogdanov37 There are 2 things here:

1. Adding to the serializer interface might be done without breaking by using default - OK, but I would still be concerned over adding more to that interface. I would like to keep it as simple as possible. So, I would be OK if this setting was outside the serializer.

2. This is a breaking change because data saved with the old code cannot be read with the new code - we don't have test for that yet, I am inferring this by looking at the code. If you can prove that data saved with the old code is read with the new, that would close this concern.


If not, I plan to address this in the next milestone. I am not 100% satisfied with our current serializer approach.

Reply:
Thanks for replying. I can prove it. It's not a problem. Current tests for load base64 encoded string works fine. 
I implement my idea with extend interface. If you want to look at I can provide new mr. Let me now if you want.

Reply:
Yep. I wrote about it. If you want I double check it.

Reply:
@abogdanov37 Just confirming. Thanks. I will check that later before merging the PR. If that confirms, then this PR can be merged without any extra config.

Reply:
@artursouza I double-check PR changes. It has full backward compatibility.

Reply:
I validated that this PR is a breaking change, like I thought originally. This is the error I get:

```txt
== APP == com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `int` from String "MjE=": not a valid Integer value

== APP ==  at [Source: (byte[])""MjE=""; line: 1, column: 1]

== APP ==       at com.fasterxml.jackson.databind.exc.InvalidFormatException.from(InvalidFormatException.java:67)

== APP ==       at com.fasterxml.jackson.databind.DeserializationContext.weirdStringException(DeserializationContext.java:1702)

== APP ==       at com.fasterxml.jackson.databind.DeserializationContext.handleWeirdStringValue(DeserializationContext.java:947)

```

How to reproduce:

1. `git checkout master`
2. `mvn clean install`
3. Apply patch attached in this comment (unzip first, of course): `git apply modify_actor_example.patch`
4. Run on a separate terminal: `cd examples; dapr run --components-path ./components/actors --app-id demoactorservice --app-port 3000  -- java -jar target/dapr-java-sdk-examples-exec.jar io.dapr.examples.actors.DemoActorService -p 3000`
5. Run the client: `cd examples; dapr run --components-path ./components/actors --app-id demoactorclient -- java -jar target/dapr-java-sdk-examples-exec.jar io.dapr.examples.actors.DemoActorClient`
6. Let it run for a few seconds (10 or so). Then, stop it with Ctrl+C or something.
7. `cd ..` to go back to project's main directoy.
8. `git stash` to stash the changes from step 3.
9. `gh pr checkout 539` to apply the changes for this PR.
10. `git stash pop` to reapply changes from step 3.
11. `mvn clean install`
12. Repeat step 4.
13. Repeat step 5.
14. Now go to terminal for step 12 and see exceptions like the one above.

[modify_actor_example.zip](https://github.com/dapr/java-sdk/files/6645320/modify_actor_example.zip)

Reply:
Hi! Thanks for example. Sorry for my bad testing. I check only for objects but not for primitives (((

Reply:
I found my place in my code where I convert Base64. I can reproduce this behavior in tests. Take a time to update code to avoid breaking changes. Thank you @artursouza!

Reply:
@abogdanov37 You might want to enable the new behavior via a System property, this way, apps can opt in when they know it will not break them (new app, for example).

Reply:
```
        T response;
        try {
          response = this.stateSerializer.deserialize(s, type);
        } catch (IOException ex) {
          byte[] data = Arrays.copyOfRange(s, 1, s.length - 1);
          response = this.stateSerializer.deserialize(Base64.getDecoder().decode(data), type);
        }
```
this solution works for all types exclude String. Strings pass through as is (base64 encoded)
Feature flag is a great idea!
By System property you mean JVM property?

Reply:
@artursouza  I push another solution based on "fair" conversion. 
If we look there https://github.com/dapr/java-sdk/blob/0e5e3112de9604b874b1832111e5a453b797eaeb/sdk-actors/src/main/java/io/dapr/actors/runtime/DaprHttpClient.java#L98 we can see that we have a transformation to Base64
I add code to transform Base64 from store to origin byte array which can be processed.